### PR TITLE
make eval garbage collection be aware of rescheduling

### PR DESCRIPTION
**Pull request by preetapan**
This PR modifies the garbage collection logic on the server side to make sure that any allocations that we will need for rescheduling tracking are not garbage collected

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,7 +7,7 @@ __BACKWARDS INCOMPATIBILITIES:__
 
 IMPROVEMENTS:
  * discovery: Allow `check_restart` to be specified in the `service` stanza.
-   [GH-3718]
+   [[GH-3718](https://github.com/hashicorp/nomad/issues/3718)]
```
Comment (chelseakomlo) [inline CHANGELOG.md]: Add description and subsystem this PR is for.

File: nomad/core_sched.go [hunk 2]
```diff
@@ -241,16 +241,18 @@ func (c *CoreScheduler) gcEval(eval *structs.Evaluation, thresholdIndex uint64,
 	// Create a watchset
 	ws := memdb.NewWatchSet()
 
+	// Look up the job
+	job, err := c.snap.JobByID(ws, eval.Namespace, eval.JobID)
```
Comment (chelseakomlo) [inline nomad/core_sched.go]: Will c.snap ever be nil?

File: nomad/core_sched_test.go [hunk 3]
```diff
@@ -209,21 +317,27 @@ func TestCoreScheduler_EvalGC_BatchStopped(t *testing.T) {
 	job := mock.Job()
 	job.Type = structs.JobTypeBatch
 	job.Status = structs.JobStatusDead
+	job.Stop = true
+	job.TaskGroups[0].ReschedulePolicy = &structs.ReschedulePolicy{
+		Attempts: 0,
+		Interval: 0 * time.Second,
```
Comment (chelseakomlo) [inline nomad/core_sched_test.go]: Can this just be 0?

File: nomad/structs/structs.go [hunk 4]
```diff
@@ -5179,6 +5179,33 @@ func (a *Allocation) RescheduleEligible(reschedulePolicy *ReschedulePolicy, fail
...
+	}
+
+	// Most recent reschedule attempt is within time interval
+	interval := reschedulePolicy.Interval
+	lastIndex := len(a.RescheduleTracker.Events)
+	lastRescheduleEvent := a.RescheduleTracker.Events[lastIndex-1]
+	timeDiff := gcTime.UTC().UnixNano() - lastRescheduleEvent.RescheduleTime
+
+	return timeDiff > interval.Nanoseconds()
+
```
Comment (chelseakomlo) [inline nomad/structs/structs.go]: Remove empty line

File: nomad/core_sched.go [hunk 5]
```diff
@@ -286,7 +288,13 @@ func (c *CoreScheduler) gcEval(eval *structs.Evaluation, thresholdIndex uint64,
...
 	var gcAllocIDs []string
 	for _, alloc := range allocs {
-		if !alloc.TerminalStatus() || alloc.ModifyIndex > thresholdIndex {
+		var reschedulePolicy *structs.ReschedulePolicy
+		tg := job.LookupTaskGroup(alloc.TaskGroup)
+
+		if tg != nil {
+			reschedulePolicy = tg.ReschedulePolicy
+		}
+		if !alloc.GCEligible(reschedulePolicy, time.Now(), thresholdIndex) {
```
Comment (chelseakomlo) [inline nomad/core_sched.go]: Is there a reason why the current time needs to be injected as a parameter?

Comment (preetapan) [reply nomad/core_sched.go]: for easier mocking in tests, and in the future we might want to replace current time with the eval's last modified time or something like that.

File: CHANGELOG.md (hunk 1 above)
Comment (preetapan) [reply CHANGELOG.md]: I am going to back this out. This wasn't my change, make dev linkifies the changelog and it must have been missing a link.

File: nomad/core_sched.go (hunk 2 above)
Comment (preetapan) [reply nomad/core_sched.go]: It is set in the constructor by the worker, should be non nil.

File: nomad/core_sched_test.go (hunk 3 above)
Comment (preetapan) [reply nomad/core_sched_test.go]: 0 will work. I added the time.Minute for visual clarity so I know when quickly scanning the test that interval takes a duration.

File: nomad/structs/structs.go [hunk 6]
```diff
@@ -5184,6 +5184,31 @@ func (a *Allocation) RescheduleEligible(reschedulePolicy *ReschedulePolicy, fail
 	return attempted < attempts
 }
 
+// GCEligible returns if the allocation is eligible to be garbage collected
+// according to its terminal status and its reschedule trackers
+func (a *Allocation) GCEligible(reschedulePolicy *ReschedulePolicy, gcTime time.Time, thresholdIndex uint64) bool {
```
Comment (dadgar) [inline nomad/structs/structs.go]: I would prefer if this was just a helper function in the core_gc file or core_gc_helpers.go. It is super specific to that use case.

Comment (dadgar) [inline nomad/structs/structs.go]: Can also GC early if the next alloc is set and this alloc is terminal

File: nomad/core_sched.go (hunk 5 above)
Comment (dadgar) [inline nomad/core_sched.go]: You may have to apply the same trick as: https://github.com/hashicorp/nomad/blob/9c790214f9b3cb333090c2a3e119453eb634e74b/nomad/core_sched.go#L254-L277 such that a dead/stopped/removed job allows the alloc to disappear.

File: nomad/core_sched.go [hunk 7]
```diff
@@ -286,7 +288,20 @@ func (c *CoreScheduler) gcEval(eval *structs.Evaluation, thresholdIndex uint64,
 	gcEval := true
 	var gcAllocIDs []string
 	for _, alloc := range allocs {
-		if !alloc.TerminalStatus() || alloc.ModifyIndex > thresholdIndex {
+		if job == nil || job.Stop {
```
Comment (preetapan) [inline nomad/core_sched.go]: @dadgar this part needs another look - here is where I addressed your comment about nil/stopped jobs, with a clarifying comment on why we should not consider a dead status job as GC eligibile.

File: nomad/state/state_store.go [hunk 8]
```diff
@@ -1889,6 +1889,21 @@ func (s *StateStore) upsertAllocsImpl(index uint64, allocs []*structs.Allocation
 			return fmt.Errorf("alloc insert failed: %v", err)
 		}
 
+		if alloc.PreviousAllocation != "" {
```
Comment (preetapan) [inline nomad/state/state_store.go]: This needs a rebase, will fix

File: nomad/core_sched.go [hunk 9]
```diff
@@ -241,31 +241,24 @@ func (c *CoreScheduler) gcEval(eval *structs.Evaluation, thresholdIndex uint64,
...
-		if job == nil {
-			collect = true
-		} else if job.Status != structs.JobStatusDead {
-			collect = false
-		} else if job.Stop {
-			collect = true
-		} else if allowBatch {
-			collect = true
-		}
+		collect := shouldCollect(job, allowBatch)
```
Comment (dadgar) [inline nomad/core_sched.go]: Hmm, this is a super generic name. Why even factor it out if it is only being called once?

File: nomad/core_sched.go [hunk 10]
```diff
@@ -286,7 +279,18 @@ func (c *CoreScheduler) gcEval(eval *structs.Evaluation, thresholdIndex uint64,
 	gcEval := true
 	var gcAllocIDs []string
 	for _, alloc := range allocs {
-		if !alloc.TerminalStatus() || alloc.ModifyIndex > thresholdIndex {
+		if job == nil || job.Stop || job.Status == structs.JobStatusDead {
```
Comment (dadgar) [inline nomad/core_sched.go]: You still need to ensure that the alloc is terminal and modify index is less than threshold

File: nomad/core_sched.go [hunk 11]
```diff
@@ -559,3 +578,32 @@ func (c *CoreScheduler) partitionDeploymentReap(deployments []string) []*structs
 
 	return requests
 }
+
+// gcEligible returns if the allocation is eligible to be garbage collected
+// according to its terminal status and its reschedule trackers
+func gcEligible(a *structs.Allocation, reschedulePolicy *structs.ReschedulePolicy, gcTime time.Time, thresholdIndex uint64) bool {
```
Comment (dadgar) [inline nomad/core_sched.go]: Can you change this to allocGcEligible

File: nomad/core_sched.go (hunk 10 above)
Comment (dadgar) [reply nomad/core_sched.go]: May want to move the logic into the gcEligible function

Comment (preetapan) [reply nomad/core_sched.go]: When the job is stopped or dead wouldn't there be no allocs that are non terminal?

Comment (dadgar) [reply nomad/core_sched.go]: Not necessarily. The scheduler still needs to run and mark the allocations as desired state stop.

**Comment by dadgar** [comment]
LGTM but you have a test failure: 

--- FAIL: TestAllocation_GCEligible (0.00s)
    --- FAIL: TestAllocation_GCEligible/GC_when_next_alloc_id_is_set (0.00s)
    	core_sched_test.go:1893: expected true but got false

**Pull request merged**